### PR TITLE
Fix key lookup in key manager

### PR DESCRIPTION
**Pull request by cydanil**
It turns out that the actual value returned by the database query does not matter.

What is important, is to know whether or not there exist a matching ingredient in the database.  
By doing so, we can establish a baseline score for the given ingredient value.
Further tests are done on variations of that spelling (notably the plural form) to maximise that accuracy.

This machinery is used to propose auto-completion when a user starts typing in an ingredient. 
I figure this is used in order to reduce the spelling of ingredients.

I'm still learning how the `ingredient` table works. Once I'll feel confident, I indent to make a PR documenting that part of the database schema.

### CODE REVIEW

File: gourmet/keymanager.py [hunk 1]
```diff
@@ -142,16 +142,20 @@ def look_for_key (self, txt):
         The higher the score, the more probable the match.
         """
 
-        txt = txt.lower()
+        txt = txt.casefold()
         retvals = {}
```
Comment (maweki) [inline gourmet/keymanager.py]: I know this is not what you looked at, but do we want to make this a `defaultdict(float)`? Then we would remove the lines down below that set `0` for every value not in `retvals`. Might be more readable for the next guy and that might be you :D

Comment (cydanil) [reply gourmet/keymanager.py]: Makes sense!

**Comment by cydanil** [comment]
I've also removed an unused import, couldn't resist

**Review by maweki — APPROVED** [review]
I think this looks good and the additional touch-up is an improvement on readability overall. I think you can merge this at your discretion.

File: gourmet/keymanager.py [hunk 2]
```diff
@@ -142,24 +143,26 @@ def look_for_key (self, txt):
...
+            # Look up for an entry in the ingredient table where the term t is
+            # found in the value from the ingkey column.
+            # By doing so, it establishes a baseline for the accuracy of t
+            # being a useful keyword.
+            entry = self.rm.fetch_one(self.rm.ingredients_table, ingkey=t)
+            if entry is not None:
                 retvals[t]=.9
             exact = self.rm.fetch_all(self.rm.keylookup_table,
                                       item=t)
             if exact:
```
Comment (maweki) [inline gourmet/keymanager.py]: Does exact always return something iterable or sometimes `None`? Maybe we can remove the `if` as well.

**Comment by cydanil** [comment]
Thank you! This is leading up to fixing the `RecCard`.

I forgot to mention that originally, the `if is_key >= 0` statement was broken, which led me to these changes.  
Indeed, what's returned by sqlalchemy is not the table's primary key as an integer, but rather the whole row as a tuple.

File: gourmet/keymanager.py (hunk 2 above)
Comment (cydanil) [reply gourmet/keymanager.py]: Yes, `exact` is always a `list`, the if statement is redundant.

**Comment by maweki** [comment]
https://github.com/kirienko/gourmet/blob/1c5a5182eb284b5f2eeba49541913d864c86d93a/gourmet/backends/db.py#L734-L737

according to https://docs.python.org/3.8/library/sqlite3.html#sqlite3.Cursor.fetchall "An empty list is returned when no rows are available."

So we can leave the `if` out as well and in that case would just iterate over the empty list.

**Comment by cydanil** [comment]
I went heavy-handed on the cosmetic changes, I would appreciate another look

File: gourmet/keymanager.py [hunk 3]
```diff
@@ -135,55 +136,64 @@ def get_key (self,txt, certainty=0.61):
         debug("End get_key",10)
         return k
 
-    def look_for_key (self, txt):
-        """handed a key, return a sorted ALIST of potential keys and
-        scores.
+    def look_for_key(self, txt: str) -> list:
```
Comment (maweki) [inline gourmet/keymanager.py]: We know what kind of list it is. It's a list of pairs with `str` and `float`

File: gourmet/keymanager.py [hunk 4]
```diff
@@ -194,12 +204,16 @@ def look_for_key (self, txt):
                 # resulted in this key, matches is the number of keys
                 # that match this word in all, and words is the number
                 # of words we're dealing with.
+                ik = match.ingkey
                 words_in_key = len(ik.split())
-                if words_in_key > nwords: wordcount = words_in_key
-                else: wordcount = nwords
-                retvals[ik]+=(float(m.count)/total_count)*(float(1)/(wordcount))
+                wordcount = words_in_key if words_in_key > nwords else nwords
+                retvals[ik] += ((float(match.count) / total_count)
```
Comment (maweki) [inline gourmet/keymanager.py]: we can remove the `float` as the division is always float in python3

File: gourmet/keymanager.py [hunk 5]
```diff
@@ -194,12 +204,16 @@ def look_for_key (self, txt):
...
+                                * (1.0 / wordcount))
+
                 # Add some probability if our word shows up in the key
-                if ik.find(w)>=0: retvals[ik]+=0.1
+                if ik.find(word) >= 0:
```
Comment (maweki) [inline gourmet/keymanager.py]: this is `word in ik`

**Review by maweki — CHANGES_REQUESTED** [review]

File: gourmet/keymanager.py (hunk 3 above)
Comment (cydanil) [reply gourmet/keymanager.py]: I've updated it

File: gourmet/keymanager.py [hunk 6]
```diff
@@ -135,55 +136,64 @@ def get_key (self,txt, certainty=0.61):
         debug("End get_key",10)
         return k
 
-    def look_for_key (self, txt):
-        """handed a key, return a sorted ALIST of potential keys and
-        scores.
+    def look_for_key(self, txt: str) -> list((str, float)):
```
Comment (maweki) [inline gourmet/keymanager.py]: https://docs.python.org/3/library/typing.html#typing.List
I think the typing hint for List needs to be different.
For Tuple as well https://docs.python.org/3/library/typing.html#typing.Tuple

so `List(Tuple(str, float))`

Comment (cydanil) [reply gourmet/keymanager.py]: Thanks for pointing it out, I'll refresh my memory.

**Comment by maweki** [comment]
Great Fix-up :)

**Comment by cydanil** [comment]
Cool! I've made an issue to create the relevant tests in another PR

**Pull request merged**